charger/phoenix-charx.go

```diff
@@ -0,0 +1,278 @@
+		return 0, err
+	}
+
+	return encoding.Uint16(b), nil
+}
+
+// Status implements the api.Charger interface
+func (wb *PhoenixCharx) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegStatus), 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	// TODO check IEC 61851-1 C1 state
+	state := string(b[0])
+
+	return api.ChargeStatus(state), nil
+}
+
+// Enabled implements the api.Charger interface
+func (wb *PhoenixCharx) Enabled() (bool, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegEnable), 1)
+	if err != nil {
+		return false, err
+	}
+
+	return encoding.Uint16(b) == 1, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *PhoenixCharx) Enable(enable bool) error {
+	b := make([]byte, 2)
+	if enable {
+		binary.BigEndian.PutUint16(b, 1)
+	}
+
+	_, err := wb.conn.WriteMultipleRegisters(wb.register(charxRegEnable), 1, b)
+
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *PhoenixCharx) MaxCurrent(current int64) error {
+	if current < 6 {
+		return fmt.Errorf("invalid current %d", current)
+	}
+
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, uint16(current))
+
+	_, err := wb.conn.WriteMultipleRegisters(wb.register(charxRegMaxCurrent), 1, b)
+
+	return err
+}
+
+var _ api.ChargeTimer = (*PhoenixCharx)(nil)
+
+// ChargingTime implements the api.ChargeTimer interface
+func (wb *PhoenixCharx) ChargingTime() (time.Duration, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegChargeTime), 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return time.Duration(encoding.Uint16(b)) * time.Second, nil
+}
+
+// currentPower implements the api.Meter interface
+func (wb *PhoenixCharx) currentPower() (float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegPower), 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Int32(b)) / 1e3, nil
+}
+
+// totalEnergy implements the api.MeterEnergy interface
+func (wb *PhoenixCharx) totalEnergy() (float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegEnergy), 4)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Int64(b)) / 1e3, nil
+}
+
+// currents implements the api.PhaseCurrents interface
+func (wb *PhoenixCharx) currents() (float64, float64, float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegCurrents), 3*2)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	return float64(encoding.Int32(b)) / 1e3,
+		float64(encoding.Int32(b[4:])) / 1e3,
+		float64(encoding.Int32(b[8:])) / 1e3, nil
+}
+
+// voltages implements the api.PhaseVoltages interface
+func (wb *PhoenixCharx) voltages() (float64, float64, float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(wb.register(charxRegVoltages), 3*2)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	return float64(encoding.Int32(b)) / 1e3,
+		float64(encoding.Int32(b[4:])) / 1e3,
+		float64(encoding.Int32(b[8:])) / 1e3, nil
+}
```

Comment:
Na die kannste aber auch noch in eine gemeinsame Funktion verpacken und für die Einzelwerte eine schöne Schleife formen.

Reply:
Ja sicher. Aber dann brauch ich ein Array (Speicherallokation) und Funktionsaufrufe.